### ISSUE TITLE
Incorrect comment assigned to function expression

### ISSUE THREAD

**Issue by tschaub** [eslint/eslint#4900]
With the change in 0d50df705f1df2d1486c7fdad14ab0aefd31215f, an incorrect comment block may be assigned to a function expression.

This used to pass `valid-jsdoc`, now it generates `Missing JSDoc @returns for function`:

```js
/* eslint valid-jsdoc: [2, {requireReturn: false}] */

/**
 * Create a new thing.
 */
var thing = new Thing({
  foo: function() {
    return 'bar';
  }
});
```

**Comment by eslintbot** [comment]
@tschaub Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by nzakas** [comment]
And this was working before that commit was merged?

**Comment by nzakas** [comment]
I should say, what did it do before? (Please note the instructions in https://github.com/eslint/eslint/issues/4900#issuecomment-170288874 for reporting a bug

**Comment by tschaub** [comment]
> 1. The version of ESLint you are using (run eslint -v)

This looks like a regression introduced by 0d50df705f1df2d1486c7fdad14ab0aefd31215f

> 2. What you did (the source code and ESLint configuration)

Here is the source code and complete config (same as in the original description):
```
/* eslint valid-jsdoc: [2, {requireReturn: false}] */

/**
 * Create a new thing.
 */
var thing = new Thing({
  foo: function() {
    return 'bar';
  }
});
```

> 3. The actual ESLint output complete with numbers

```
$ git checkout 0d50df705f1df2d1486c7fdad14ab0aefd31215f
$ node bin/eslint.js --no-eslintrc script.js

/Users/tschaub/projects/eslint/script.js
  3:1  error  Missing JSDoc @returns for function  
```

> 4. What you expected to happen instead

I would expect the above code to be valid according to `valid-jsdoc`.  The documentation block is not associated with the function expression.

Prior to 0d50df705f1df2d1486c7fdad14ab0aefd31215f, the `valid-jsdoc` rule treated this code as valid (as it should):

```
$ git checkout 0d50df705f1df2d1486c7fdad14ab0aefd31215f^
$ node bin/eslint.js --no-eslintrc script.js
$
```

The issue is that the `sourceCode.getJSDocComment()` traverses up to the assignment expression, finds the leading comments, and returns these as the comment associated with the function expression.

**Comment by tschaub** [comment]
@nzakas let me know if the description above is still not clear.  The change in #4926 corrects this particular case.
